# Tasks


#### 1: If you haven’t done so already, perform the wrangling procedures you walked through in this Exercise on your project data in a new notebook for this Exercise. Then, add a new section heading to separate your wrangling procedures from the procedures you’ll be conducting in the steps below.

#### 2: Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

#### 3: Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.
    
#### 4: Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.
    
#### 5: Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.
    
#### 6: The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.
    
#### 7: They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.
    
#### 8: It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?
    
#### 9: Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.
    
#### 10: You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?
    
#### 11: Check the organization and structure of your notebook. Be sure to include section headings and code comments.
    
#### 12: Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.
    
#### 13: Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.
    
#### 14: Save your Jupyter notebook and submit it here for your tutor to review.


## 01. Import Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## Importing Data

In [2]:
path = r'C:\Users\maxca\Instacart Basket Analysis'

In [3]:
# Order Data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [4]:
# Product Data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

## Data Wrangling

#### 1: If you haven’t done so already, perform the wrangling procedures you walked through in this Exercise on your project data in a new notebook for this Exercise. Then, add a new section heading to separate your wrangling procedures from the procedures you’ll be conducting in the steps below.

## Dropping Columns

In [5]:
#Dropping eval_set from Orders >just visual<
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [17]:
# Actually removing eval_set >always be careful - this can't be undone<
df_ords = df_ords.drop(columns = ['eval_set'])

In [18]:
# Better way of dropping data is to create and rename the df into another one, which makes mistakes way more manageable -> space problems
df_ords_2 = df_ords.drop(columns = ['eval_set']) #doesn't work anymore since eval_set is already dropped

KeyError: "['eval_set'] not found in axis"

## Spotting missing  values

In [ ]:
# Spotting missing values with the df['variable'].value_counts(dropna = False) command here on days_since_prior_order
df_ords['days_since_prior_order'].value_counts(dropna = False)


## Renaming Columns

In [46]:
# renaming a column with the command df.rename(columns = {'old_name' : 'new_name'}, inplace = True)
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [47]:
df_ords.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

## Changeing Variable's Data Type

In [19]:
# You don't always want to df.describe() all the data. Example: order_id is just a key. To exclude change the data type so
# df.describe ignores it
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [20]:
df_ords['order_id'].dtype

dtype('O')

## Transposing Data

In [21]:
# Importing Departments data 
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [22]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [23]:
# Transposing df_dep
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [24]:
# New data frame df_dep_t
df_dep_t = df_dep.T

In [25]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [26]:
#add an index
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [27]:
# new header for row 0
new_header = df_dep_t.iloc[0]

In [28]:
new_header 

0    department
Name: department_id, dtype: object

In [29]:
# delete the old row 0 by copying everyting starting with row 1
df_dep_t_new = df_dep_t[1:]

In [30]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [31]:
# now place our new_header as the header of df_dep_t_new
df_dep_t_new.columns = new_header

In [32]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

## Creating a Data Dictionarie

In [33]:
# command to turn df_dep_t_new into a data dictionarie
data_dict = df_dep_t_new.to_dict('index')

In [34]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [35]:
# Using the dictionarie
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [36]:
# check what department_id 19 stands for
print(data_dict.get('19'))

{'department': 'snacks'}


## Subsetting

In [37]:
# create subset only containing data from the snacks department
df_snacks = df_prods[df_prods['department_id']==19]

In [38]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [39]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

In [40]:
# easier:
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

In [41]:
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

## Exercise

#### 2: Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [75]:
df_ords.head()

order_id user_id  number_of_orders  orders_day_of_week  order_hour_of_day  \
0  2539329       1                 1                   2                  8   
1  2398795       1                 2                   3                  7   
2   473747       1                 3                   3                 12   
3  2254736       1                 4                   4                  7   
4   431534       1                 5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [45]:
# 3 user_id doesn't have to be a numeric value

In [ ]:
df_ords['user_id'] = df_ords['user_id'].astype('str')

#### 3: Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [77]:
df_ords.columns

Index(['order_id', 'user_id', 'number_of_orders', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [51]:
# order_number should be number_of_orders

In [ ]:
df_ords.rename(columns = {'order_number' : 'number_of_orders'}, inplace = True)

In [52]:
df_ords.head()

order_id user_id  number_of_orders  orders_day_of_week  order_hour_of_day  \
0  2539329       1                 1                   2                  8   
1  2398795       1                 2                   3                  7   
2   473747       1                 3                   3                 12   
3  2254736       1                 4                   4                  7   
4   431534       1                 5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

#### 4: Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [79]:
df_ords['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

In [ ]:
# Answer: 10 AM

#### 5: Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.

In [81]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [57]:
print(data_dict.get('4'))

{'department': 'produce'}


#### 6: The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [83]:
df_prods[df_prods['department_id']==14]

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [84]:
df_breakfast =  df_prods[df_prods['department_id']==14]

In [85]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

#### 7: They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [ ]:
df_dinner_party = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [62]:
df_dinner_party

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

##### 8: It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?
#### Answer: df_dinner_party with 7650 rows.

#### 9: Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [69]:
user_id_1 = df_ords[df_ords['user_id']=='1']

In [70]:
user_id_1

order_id user_id  number_of_orders  orders_day_of_week  order_hour_of_day  \
0   2539329       1                 1                   2                  8   
1   2398795       1                 2                   3                  7   
2    473747       1                 3                   3                 12   
3   2254736       1                 4                   4                  7   
4    431534       1                 5                   4                 15   
5   3367565       1                 6                   2                  7   
6    550135       1                 7                   1                  9   
7   3108588       1                 8                   1                 14   
8   2295261       1                 9                   1                 16   
9   2550362       1                10                   4                  8   
10  1187899       1                11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

#### 10: You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

In [87]:
user_id_1.describe()

number_of_orders  orders_day_of_week  order_hour_of_day  \
count         11.000000           11.000000          11.000000   
mean           6.000000            2.636364          10.090909   
std            3.316625            1.286291           3.477198   
min            1.000000            1.000000           7.000000   
25%            3.500000            1.500000           7.500000   
50%            6.000000            3.000000           8.000000   
75%            8.500000            4.000000          13.000000   
max           11.000000            4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

#### 12: Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

#### 13: Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format.

In [ ]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))